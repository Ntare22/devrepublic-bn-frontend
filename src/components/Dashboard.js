import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import NavBar from './sharedComponents/AuthNavBar';

const Dashboard = ({ token, image }) => (<NavBar token={token} image={image} />);

Dashboard.propTypes = {
  token: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};

const mapStateToProps = ({ user }) => ({
  token: user.token || localStorage.getItem('token'),
  image: '',
});

export default connect(mapStateToProps)(Dashboard);
