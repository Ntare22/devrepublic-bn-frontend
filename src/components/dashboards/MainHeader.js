import React from 'react';
import { Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import MainPanel from '../sharedComponents/MainPanel';

const MainHeader = ({ filter, allTrips }) => (
  <Col className="requester-Dashboard">
    <div className="request-dashboard-title ">
      <MainPanel allTrips={allTrips} filter={filter} />
    </div>
  </Col>
);

MainHeader.propTypes = {
  filter: PropTypes.func,
  allTrips: PropTypes.array,
};

export default MainHeader;
