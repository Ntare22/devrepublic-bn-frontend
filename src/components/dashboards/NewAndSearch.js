import React from 'react';
import '../../scss/requestTable.scss';
import searchIcon from '../../assets/search-icon.png';
import addIcon from '../../assets/add-icon.png';

const NewAndSearch = () => (
  <div>
    <img className="icon" src={searchIcon} alt="search icon" />
    <img className="icon add-icon" src={addIcon} alt="add icon" />
  </div>
);

export default NewAndSearch;
