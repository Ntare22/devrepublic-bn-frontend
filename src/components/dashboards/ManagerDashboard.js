import React, { Component } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Table from 'react-bootstrap/Table';
import Container from 'react-bootstrap/Container';
import '../../scss/requestTable.scss';
import AuthNavBar from '../sharedComponents/AuthNavBar';
import Footer from '../sharedComponents/Footer';
import { getRequestsAction } from '../../redux/actions/getRequests';
import {
  checkStatus, requestDates, columnTitles, reasonColumnTitle, spinnner, dashboardSidebar,
} from './sharedFuncs';
import NewAndSearch from './NewAndSearch';
import placeholder from '../../assets/users.png';
import translate from '../languages/Translate';
import pagination from '../../util/pagination';
import MainHeader from './MainHeader';


class ManagerDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      requests: [],
      displayedRequests: [],
      loading: true,
    };
  }

  async componentDidMount() {
    const { token } = this.props;
    const actionOutput = await this.props.getRequestsAction(token);
    this.setState({
      requests: actionOutput.payload.data.requests,
      loading: false,
    });
    this.setState((prevState) => ({ displayedRequests: pagination(prevState.requests, 1) }));
  }

  filter = (newArray) => {
    this.setState({
      displayedRequests: newArray,
    });
  }

  render() {
    const { requests, loading, displayedRequests } = this.state;
    const requestsLength = requests.length;
    const Requester = 'Requester';
    const tripDetails = 'Trip Details';
    const date = 'Date';
    const Status = 'Status';

    return (
      <>
        <AuthNavBar />
        <div>
          <Container className="whole-conatainer" fluid>
            <Row className="bg-dange dashboard-view">
              <Col lg="2" sm="1" xs="1" md="2">
                {dashboardSidebar()}

              </Col>
              <Col lg="10" sm="11" md="9" xs="11" className="col-1">

                <Row>
                  <MainHeader allTrips={requests} filter={this.filter} />
                  <Col className="all-icons text-center">
                    <NewAndSearch />
                  </Col>

                </Row>
                {
                  loading ? (
                    <>
                      {spinnner()}
                    </>
                  ) : ''
                }
                {
                                              requestsLength === 0 && !loading ? (
                                                <div className="no-requests mt-2">
                                                  {translate('No request created by your direct reports yet!')}
                                                </div>
                                              ) : (
                                                <div className="table-header">
                                                  <Table>
                                                    {columnTitles(Requester, tripDetails, date, Status)}
                                                    {displayedRequests.map((request) => (
                                                      <tbody key={request.id}>
                                                        <tr className="row-info-container">
                                                          <td className="col-accomodation">
                                                            <span>
                                                              <img className="profile-image" src={request.User.image || placeholder} alt="profile" />
                                                            </span>
                                                            <span className="firstName">{request.User.firstName}</span>
                                                            <span className="lastName">{request.User.lastName}</span>

                                                          </td>
                                                          {reasonColumnTitle(request)}

                                                          {requestDates(request)}
                                                          <td className="text-center col-status">
                                                            {checkStatus(request)}
                                                          </td>
                                                        </tr>
                                                      </tbody>

                                                    ))}
                                                  </Table>
                                                </div>
                                              )
  }
                <Row />
              </Col>
            </Row>
          </Container>
        </div>
        <Footer />
      </>
    );
  }
}

const MapStateToProps = ({ requests }) => ({
  requests,
});
export { ManagerDashboard };
export default connect(MapStateToProps, { getRequestsAction })(ManagerDashboard);

ManagerDashboard.propTypes = {
  getRequestsAction: PropTypes.func,
  token: PropTypes.string,
};
