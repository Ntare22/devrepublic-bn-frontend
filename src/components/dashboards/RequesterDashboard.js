import React, { Component } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import { connect } from 'react-redux';
import Table from 'react-bootstrap/Table';
import PropTypes from 'prop-types';
import Container from 'react-bootstrap/Container';
import '../../scss/requestTable.scss';
import AuthNavBar from '../sharedComponents/AuthNavBar';
import Footer from '../sharedComponents/Footer';
import NewAndSearch from './NewAndSearch';
import { getRequestsAction } from '../../redux/actions/getRequests';
import {
  checkStatus, columnTitles, reasonColumnTitle, spinnner, dashboardSidebar,
}
  from './sharedFuncs';
import translate from '../languages/Translate';

class RequesterDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      requests: [],
      bookings: [],
      loading: true,
    };
  }

  async componentDidMount() {
    const { token } = this.props;
    const actionOutput = await this.props.getRequestsAction(token);
    this.setState({
      requests: actionOutput.payload.data.requests,
      bookings: actionOutput.payload.data.bookings,
      loading: false,
    });
  }

  findAccommodation = (request, booking) => {
    switch (request.id) {
      case `${booking.requestId}`:
        return (
          <p className="ml-1 bg-danger">{booking.facilityName}</p>
        );
      default:
        return (
          <p className="tex-center ml-5">-</p>
        );
    }
  }

  render() {
    const { requests, bookings, loading } = this.state;
    const requestsLength = requests.length;
    const Accommodation = 'Accommodation';
    const tripDetails = 'Trip Details';
    const date = 'Departure Date';
    const Status = 'Status';
    return (
      <>
        <AuthNavBar />
        <div>
          <Container className="whole-conatainer" fluid>
            <Row>
              <Col lg="2" sm="1" xs="1" md="2">
                {dashboardSidebar()}
              </Col>
              <Col lg="10" sm="11" md="9" xs="11" className="col-1">

                <Row>
                  <Col className="requester-Dashboard">
                    <div className="request-dashboard-title ">
                      { translate('Requester Dashboard') }
                    </div>
                  </Col>
                  <Col className="all-icons text-center ">
                    <div>
                      <NewAndSearch />
                    </div>
                  </Col>
                </Row>
                {
                  loading ? (
                    <>
                      {spinnner()}
                    </>
                  ) : ''
                }
                {
                            requestsLength === 0 && !loading ? (
                              <div className="no-requests mt-2">
                                {translate('No request found. Create a request!')}
                              </div>
                            ) : (
                              <div className="dashboard-content">
                                <Table>
                                  {columnTitles(tripDetails, Accommodation, date, Status)}
                                  {requests.map((request) => (
                                    <tbody key={request.id} className="table-header">
                                      <tr className="row-info-container">
                                        {reasonColumnTitle(request)}
                                        <td className="col-accomodation">
                                          {
                              bookings.map((booking) => (
                                this.findAccommodation(request, booking)
                              ))
                            }
                                        </td>
                                        <td className="col-departure-date">{request.departureDate}</td>
                                        <td className="text-center col-status">
                                          { checkStatus(request) }
                                        </td>
                                      </tr>
                                    </tbody>

                                  ))}
                                </Table>
                              </div>
                            )
                }
                <Row />
              </Col>
            </Row>
          </Container>
        </div>
        <Footer />
      </>
    );
  }
}

const MapStateToProps = ({ requests }) => ({
  requests,
});
export { RequesterDashboard };
export default connect(MapStateToProps, { getRequestsAction })(RequesterDashboard);

RequesterDashboard.propTypes = {
  getRequestsAction: PropTypes.func,
  token: PropTypes.string,
};
