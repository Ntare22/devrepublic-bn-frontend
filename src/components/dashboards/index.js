import React, { Component } from 'react';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import jwtDecode from 'jwt-decode';
import RequesterDashboard from './RequesterDashboard';
import ManagerDashboard from './ManagerDashboard';
import AdminDashboard from './admin';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      role: '',
      token: '',
    };
  }

  componentDidMount() {
    const { token } = queryString.parse(this.props.location.search);
    const localStorageToken = localStorage.getItem('token');
    const userToken = localStorageToken || token;
    const { role } = jwtDecode(userToken);
    this.setState({ role, token: userToken });
  }

  checkRole = (role) => {
    const { token } = this.state;
    switch (role) {
      case 'requester':
        return <RequesterDashboard token={token} history={this.props.history} />;
      case 'manager':
        return <ManagerDashboard token={token} />;
      case 'super administrator':
        return <AdminDashboard />;
      default:
        return null;
    }
  }

  render() {
    const { role } = this.state;
    return (
      <>
        { this.checkRole(role) }
      </>
    );
  }
}

export default Dashboard;

Dashboard.propTypes = {
  location: PropTypes.object,
  history: PropTypes.object,
};
