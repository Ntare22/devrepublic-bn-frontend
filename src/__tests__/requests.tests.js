import * as React from 'react';
import { mount, shallow } from 'enzyme';
import dotenv from 'dotenv';
import toJson from 'enzyme-to-json';
import { RequesterDashboard as Requester } from '../components/dashboards/RequesterDashboard';
import { ManagerDashboard as Manager } from '../components/dashboards/ManagerDashboard';
import Dashboard from '../components/dashboards/index';
import NewAndSearch from '../components/dashboards/NewAndSearch';
import { checkStatus } from '../components/dashboards/sharedFuncs';

dotenv.config();

describe('Requester dashboard snapshot', () => {
  const render = (params, fn = shallow) => {
    const defaultProps = {
      getRequestsAction: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <Requester {...props} />,
    );
  };
  it('should create a dashboard snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
  it('should call getRequests Action and return requests', () => {
    const wrapper = render();
    wrapper.setState({
      requests: [],
    });
    const { getRequestsAction } = wrapper.instance().props;
    getRequestsAction.mockReturnValueOnce({
      payload: {
        data: {
          requests: [{ id: 'asedr45', userId: 'wert45' }],
        },
      },
    });
    wrapper.setState({
      requests: [{ id: 'asedr45', userId: 'wert45' }],
    });
    expect(wrapper.state().requests).toHaveLength(1);
    wrapper.instance().findAccommodation({ id: 'wer345d' }, { requestId: 'wer345d' });
    wrapper.instance().findAccommodation({ id: 'oppop' }, { requestId: 'wer345d' });
    checkStatus({ status: 'rejected' });
    checkStatus({ status: 'open' });
    checkStatus({ status: 'approved' });
  });
  it('should call getRequests Action and return bookings', () => {
    const wrapper = render();
    wrapper.setState({
      bookings: [],
    });
    const { getRequestsAction } = wrapper.instance().props;
    getRequestsAction.mockReturnValue({
      payload: {
        data: {
          bookings: [{ facikityName: 'asedr45', facilityId: 'wert45' }],
        },
      },
    });
    wrapper.setState({
      bookings: [{ facikityName: 'asedr45', facilityId: 'wert45' },
        { facikityName: 'asedr45', facilityId: 'wert45' },
        { facikityName: 'asedr45', facilityId: 'wert45' }],
    });
    expect(wrapper.state().bookings).toHaveLength(3);
    wrapper.instance().findAccommodation({ id: 'oppop' }, { requestId: 'wer345d' });
  });
});


describe('Manager dashboard snapshot', () => {
  const render = (params, fn = shallow) => {
    const defaultProps = {
      getRequestsAction: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <Manager {...props} />,
    );
  };
  it('should create a dashboard snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
  it('Manager can get requests and user info', () => {
    const wrapper = render();
    wrapper.setState({
      requests: [],
    });
    const { getRequestsAction } = wrapper.instance().props;
    getRequestsAction.mockReturnValueOnce({
      payload: {
        data: {
          requests: [{
            id: 'asedr45',
            userId: 'wert45',
            managerId: '23dfsd45',
            User: {
              firstName: 'Aime',
              lastName: 'Bien',
              image: 'jeanne@andela.com',
            },
          }],
        },
      },
    });
    wrapper.setState({
      requests: [{
        id: 'asedr45',
        userId: 'wert45',
        managerId: '23dfsd45',
        User: {
          firstName: 'Aime',
          lastName: 'Bien',
          image: 'jeanne@andela.com',

        },
      }],
    });
    expect(wrapper.state().requests).toHaveLength(1);
  });
});
describe('Index dashboard snapshot', () => {
  const render = (params, fn = shallow) => {
    const defaultProps = {
      getRequestsActionn: jest.fn(),
      location: {},
      search: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <Dashboard {...props} />,
    );
  };
  beforeEach(() => {
    const token = process.env.TOKEN;
    localStorage.setItem('token', token);
    const managerToken = process.env.MANAGER_TOKEN;
    localStorage.setItem('token', managerToken);
  });
  it('should create a dashboard container snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
  it('check the user status before rendering dashboard', () => {
    const wrapper = render();
    wrapper.setState({
      role: '',
    });
    wrapper.setState({
      role: 'requester',
    });
    expect(wrapper.state().role).toEqual('requester');
    wrapper.instance().checkRole({ role: 'requester' });
    wrapper.instance().checkRole({ role: 'manager' });
    wrapper.instance().checkRole({ role: '45678' });
  });
});

describe('Index dashboard snapshot', () => {
  const render = mount(<NewAndSearch />);
  it('should create a NewAndSearch snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
});
