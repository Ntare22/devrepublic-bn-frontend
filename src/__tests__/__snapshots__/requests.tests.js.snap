// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index dashboard snapshot should create a NewAndSearch snapshot 1`] = `
<NewAndSearch>
  <div>
    <img
      alt="search icon"
      className="icon"
      src="test-file-stub"
    />
    <img
      alt="add icon"
      className="icon add-icon"
      src="test-file-stub"
    />
  </div>
</NewAndSearch>
`;

exports[`Index dashboard snapshot should create a dashboard container snapshot 1`] = `undefined`;

exports[`Manager dashboard snapshot should create a dashboard snapshot 1`] = `undefined`;

exports[`Requester dashboard snapshot should create a dashboard snapshot 1`] = `undefined`;
